### ISSUE TITLE
Heatmap interpretation problem

### ISSUE THREAD

**Issue by tienn2** [mahmoodlab/CLAM#63]
Hello,

I was able to train a clam_sb model on my own dataset for the classification task of breast cancer tumor _vs_ normal tissue and then was able to generate the heatmaps of some of the test slides with the create_heatmaps.py script. The test results are reasonably good: 
![image](https://user-images.githubusercontent.com/38996274/123162424-9652c600-d43e-11eb-9b59-c26325c70bb3.png)

but when looking at the heatmaps, the attention scores aren't what I expected to be. Specifically, for tumor tissues, adipose and stromal tissue patches were given the highest attention scores whereas high density nuclei patches were given the lowest attention scores. It's like if the model learned to classify backwards: tumor regions highlighted in blue (low attention) and normal tissue in red (high attention) for a tumor slide image. Here's an example of a subregion of a heatmap for a tumor slide:
![image](https://user-images.githubusercontent.com/38996274/123167753-f77d9800-d444-11eb-9caa-891bca15eac9.png)

I was expecting the regions with higher density of nuclei to be in red and the rest in blue, but the heatmaps was showing otherwise. [Here](https://drive.google.com/drive/folders/1ceahmoo1u2Tcb29_q8q8DK-wJkEofBWi?usp=sharing) are 2 full heatmaps if you want to look at the attention scores across the whole slides. The model predictions (from fold 0) are very good for these two slides:
![image](https://user-images.githubusercontent.com/38996274/123166827-d5374a80-d443-11eb-8cdf-5b91cbaa8507.png)
but the interpretation with the heatmaps get confusing since tumor regions are in blue and normal tissue are in red.

I was wondering if you ever encountered this behavior from the heatmaps outputs and if you have an intuition on what could be the problem here.

If it could help, here's some more details about my dataset and training parameters:
Dataset label/patient distribution:
- Patient-LVL; Number of samples registered in class 0: 8
- Slide-LVL; Number of samples registered in class 0: 26
- Patient-LVL; Number of samples registered in class 1: 38
- Slide-LVL; Number of samples registered in class 1: 121

Training parameters:
- 5 folds, 70/20/10 train/valid/test splits
- I used the default training settings (learning rate, model type, etc.)

Sorry for the long post, I hope it was clear enough.
Cheers!

**Comment by fedshyvana** [comment]
Hi, while I would not completely rule out the possibility of a bug in visualization (feel free to let me know if you notice anything fishy on that front), my gut feeling is the heatmap visualization you showed demonstrate the model has not learned the right features, and got it away with it (for most splits) simply due to being validated on a small test set (presumably ~10 slides?). 

Due to the weakly-supervised nature, it is possible for the model to learn to incorrectly rely on certain features that are generalizable on larger training/evaluation cohorts - for example, if all tumor slides in your training set consistently have adipose tissue or certain patterns of stromal tissue that are not present in the negative slides - then from the perspective of the learning algorithm - it would be perfectly natural to use that as a means to stratify positive vs negative samples. Another possible example of this, could be that the pathologist always circles the tumor region and as a result the model learns to incorrectly associates pen marking on the slide with tumor (which we don't want). 

To verify if this is indeed the case, I would test your model on a larger independent cohort (e.g. maybe TCGA?) to see if performance tanks or maybe try to adjust the segmentation threshold such that adipose tissue is excluded from all WSIs during data processing to prevent the model from learning it as a predictive feature, or possibly observing that the model incorrectly classifies negative slides with adipose tissue as positive, etc.
